src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/loptimizer/heuristics/formula/JoinAwareWeightedUnboundVariableCount.java

```diff
@@ -0,0 +1,113 @@
+ * - number of joins contained in this subplan
+ * - types of joins contained in this subplan (star join, chain join, unusal join)
+ * To get unbound variables, the set of bindings in selectedPlans needs to be considered.
+ *
+ * This implementation is based on the formula (7) that is proposed in paper
+ * "Heuristics-based Query Reordering for Federated Queries in SPARQL 1.1 and SPARQL-LD"
+ */
+
+public class JoinAwareWeightedUnboundVariableCount {
+    protected static final double W_S = 1; // weight for subject variables
+    protected static final double W_P = 0.1; // weight for predicate variables
+    protected static final double W_O = 0.8; // weight for object variables
+
+    protected static final double J_Tc = 0.6; // weight for chain join
+    protected static final double J_Ts = 0.5; // weight for star join
+    protected static final double J_Tu = 1; // weight for unusual join
+
+    public static double estimate( final List<Query_Analyzer> selectedPlans, final Query_Analyzer subPlan ) {
+        // Add bound variables of all selected plans to a set
+        Set<Node> boundVariables = new HashSet<>();
+
+        // This part can be optimized if all formulas need to consider bound variables:
+        // adding to this set directly when adding a new subPlan to 'selectedPlans'
+        for ( Query_Analyzer plan : selectedPlans ) {
+            boundVariables = addBinds( plan.getSubs(), plan.getPreds(), plan.getObjs() );
+        }
+
+        return calculateCost( boundVariables, subPlan );
+    }
+
+    // Formula (7) in paper "Heuristics-based Query Reordering for Federated Queries in SPARQL 1.1 and SPARQL-LD"
+    private static double calculateCost( final Set<Node> bindings, final Query_Analyzer subPlan ) {
+        final double unboundVarsCost = getUnboundVarsCost(
+                subPlan.getSubs(),
+                subPlan.getPreds(),
+                subPlan.getObjs(),
+                bindings);
+        final double joinCost = joinsWeight(
+                subPlan.getSubs(),
+                subPlan.getPreds(),
+                subPlan.getObjs());
+        return unboundVarsCost / joinCost;
+    }
+
+    private static double getUnboundVarsCost( final List<Node> vars_s, final List<Node> vars_p,
+                                              final List<Node> vars_o, final Set<Node> boundVariables) {
+        final Set<Node> varsTotal = new HashSet<>();
+        // Calculate the number of (unique) unbound subjects, predicates and objects
+        final int totalSubs = calculateVars(vars_s, varsTotal, boundVariables);
+        final int totalObjs = calculateVars(vars_o, varsTotal, boundVariables);
+        final int totalPreds = calculateVars(vars_p, varsTotal, boundVariables);
+
+        return calculateTripleWeights(totalSubs, totalPreds, totalObjs);
+    }
+
+    /**
+     * Creates a collection with all variables that have been currently calculated
+     * and adds the bound variables to bindings
```

Comment:
It is still not clear what is meant by "variables that have been [...] calculated."

Additionally, there is nothing called `bindings` here.

Reply:
Hi @hartig , for the implementation of cost calculation (using formulas), since some details are missing in the paper so I use some methods from the corresponding [GitHub repository](https://github.com/TYannakis/SPARQL-LD-Query-Optimizer/blob/14902192461e41df34335e67f5c9998935b7ad9d/src/main/java/ics/forth/formulas/GreedyJWUVC.java#L200) of that paper. 

In particular, I directly used the implementation for this function of that repository. But after carefully reading it, I realized that the parameter "varsTotal" is not needed so I have removed it.